### PR TITLE
Spec update

### PR DESCRIPTION
**Pull request by william-vw**
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/N3/pull/117.html" title="Last updated on Feb 6, 2023, 3:11 PM UTC (79c46ab)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/N3/117/ab61812...79c46ab.html" title="Last updated on Feb 6, 2023, 3:11 PM UTC (79c46ab)">Diff</a>

### CODE REVIEW

**Review by gkellogg — APPROVED** [review]
For whatever reason, the files view only shows split screen, and it's difficult to add inline suggestions. In general, I think this is a big improvement, but may need some tweaks after being merged.

File: spec/index.html [hunk 1]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
-        a <a data-cite="RDF11-CONCEPTS#dfn-subject">subject</a>
-        and an <a data-cite="RDF11-CONCEPTS#dfn-object">object</a> resource.
-        Although unrestricted in N3, a
-        <a>resource path</a> is most often used in the
-        <a data-cite="RDF11-CONCEPTS#dfn-subject">subject</a> or
-        <a data-cite="RDF11-CONCEPTS#dfn-object">object</a> position.
-      </p>
+      <p>Similar to <a data-cite="SPARQL11-QUERY#propertypaths">SPARQL property paths</a> [[SPARQL11-QUERY]], 
+        N3 <dfn>resource paths</dfn> concisely express paths between resources, 
+        in case intermediary resources on the path are not relevant.</p>
```
Comment (TallTed) [inline spec/index.html:793]:
```suggestion
        when intermediary resources on the path are not relevant.</p>
```

File: spec/index.html [hunk 2]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
-          a relationship between Joe's mother's office's zipcode:</p>
+        <p>A <a>resource path</a> starts from a subject resource, followed by one or more predicates;
+          each predicate is separated by a <dfn>directional indicator</dfn> to follow the predicate either forward (`!`) or in reverse (`^`).
+        For example, the following example describes the city of Joe's mother's office's address:</p>
 
-        <pre data-transform="updateExample">
+        <pre class="example nohighlight" data-transform="updateExample">
             <!--
-            @base <http://example.org/#> .
+            @prefix : <http://example.org/#> .
```
Comment (TallTed) [inline spec/index.html:801]:
```suggestion
@prefix : <http://example.org/#> .
```

File: spec/index.html [hunk 3]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
 
-            :joe!:mother!:office!:zip a :Zipcode .
+            :joe!:hasMother!:hasOffice!:hasAddress :hasCity "Metropolis" .
```
Comment (TallTed) [inline spec/index.html:803]:
```suggestion
:joe!:hasMother!:hasOffice!:hasAddress :hasCity "Metropolis" .
```

File: spec/index.html [hunk 4]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
+          are not described or referenced further, so the more concise <a>resource path</a> syntax can be used.</p>
 
-        <pre data-transform="updateExample">
+        <p>The expansion of this shorthand syntax uses blank nodes to express the path between the two resources.
+          The example above is equivalent to the following (using blank node identifiers for clarity):</p>
+
+        <pre class="example nohighlight" data-transform="updateExample">
             <!--
-            @base <http://example.org/#> .
+            @prefix : <http://example.org/#> .
```
Comment (TallTed) [inline spec/index.html:815]:
```suggestion
@prefix : <http://example.org/#> .
```

File: spec/index.html [hunk 5]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
 
-            [is :zip of [is :office of [is :mother of :joe]]] a :ZipCode .
+            :joe :hasMother _:m . 
+            _:m :hasOffice _:o . 
+            _:o :hasAddress _:a . 
+            _:a :hasCity "Metropolis" .
```
Comment (TallTed) [inline spec/index.html:820]:
```suggestion
:joe :hasMother _:m . 
_:m :hasOffice _:o . 
_:o :hasAddress _:a . 
_:a :hasCity "Metropolis" .
```

File: spec/index.html [hunk 6]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
-          a someone who is Joe's grandmother:</p>
+        <p>Relations can also be followed in reverse using the reverse (`^`) indicator.
+          The following example, starting from `joe`,  follows the `hasMother` predicate to Joe's mother, 
+          and then follows the `hasMother` predicate in reverse (thus pointing to someone who has the same mother):</p>
 
-        <pre data-transform="updateExample">
+        <pre class="example nohighlight" data-transform="updateExample">
             <!--
-            @base <http://example.org/#> .
+            @prefix : <http://example.org/#> .
```
Comment (TallTed) [inline spec/index.html:833]:
```suggestion
@prefix : <http://example.org/#> .
```

File: spec/index.html [hunk 7]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
             -->
           </pre>
 
         <p>This could equally well be represented by the more verbose:</p>
 
-        <pre data-transform="updateExample">
+        <pre class="example nohighlight" data-transform="updateExample">
             <!--
-            @base <http://example.org/#> .
+            @prefix : <http://example.org/#> .
```
Comment (TallTed) [inline spec/index.html:843]:
```suggestion
@prefix : <http://example.org/#> .
```

File: spec/index.html [hunk 8]
```diff
@@ -788,106 +788,79 @@ <h3>Quoted Graphs</h3>
...
 
-             [ :mother [is :mother of :joe]] a :Person .
+            :joe :hasMother _:m.
+            _:d :hasMother _:m.
+            _:d a :Person.
```
Comment (TallTed) [inline spec/index.html:847]:
```suggestion
:joe :hasMother _:m.
_:d :hasMother _:m.
_:d a :Person.
```

File: spec/index.html (hunk 2 above)
Comment (gkellogg) [reply spec/index.html:801]: Indenting pre blocks shouldn't be necessary, as ReSpec does this automatically now.

Comment (gkellogg) [reply spec/index.html:801]: Rather, left-justifying.

**Comment by gkellogg** [comment]
Looks like a bug in the `updateExample` code which is not dropping the trailing `-->` in examples. Odd, as you can see the treatment in the `unComment` function. And, this code is used in other specs without issue.

File: spec/index.html (hunk 2 above)
Comment (TallTed) [reply spec/index.html:801]: > ReSpec does this automatically now

Ah, OK. They do keep improving!

**Comment by william-vw** [comment]
> Looks like a bug in the `updateExample` code which is not dropping the trailing `-->` in examples. Odd, as you can see the treatment in the `unComment` function. And, this code is used in other specs without issue.

Yes I noticed that. Unsure what's going wrong here.

**Comment by gkellogg** [comment]
If not in this/each PR, then eventually, the Rake task in /ns should be run to re-generate the .html files when the base .n3 files are updated.

**Comment by william-vw** [comment]
Hi Gregg

Looks like this <https://github.com/w3c/N3/commit/fb0f698f2b5f7218371781173ac02d12385d9b44#diff-97efdc0e30285fae4d5cb7cc4a510dbaa7c33f4e56d4216dcde109ad2cdef15fL97> was the culprit … (Maybe a replace-all that went awry)

We're in the process of automating the building of the HTML file that Tim showed today. If that all goes well, would you be able to help us with setting up a GitHub workflow to auto-generate (and possibly preview) the file?


Thanks

William

**Comment by gkellogg** [comment]
> Hi Gregg Looks like this <[fb0f698#diff-97efdc0e30285fae4d5cb7cc4a510dbaa7c33f4e56d4216dcde109ad2cdef15fL97](https://github.com/w3c/N3/commit/fb0f698f2b5f7218371781173ac02d12385d9b44#diff-97efdc0e30285fae4d5cb7cc4a510dbaa7c33f4e56d4216dcde109ad2cdef15fL97)> was the culprit … (Maybe a replace-all that went awry) We're in the process of automating the building of the HTML file that Tim showed today. If that all goes well, would you be able to help us with setting up a GitHub workflow to auto-generate (and possibly preview) the file? Thanks William
> […](#)

Sure, in theory, we can build these things in GH Actions and update the commit with changed files, so no human intervention necessary. I've not really figured this out, though.

File: spec/index.html [hunk 9]
```diff
@@ -1191,15 +1191,15 @@ <h5>Scoped Negation as Failure (SNAF)</h5>
 
       <h5>Access Online Knowledge</h5>
       <p>
-        To dynamically retrieve online knowledge within a <a>N3 rule</a>, the <code>log:content</code>,
-        <code>log:semantics</code>, and <code>log:conclusion</code> builtins allow pulling in, parsing, and reasoning
+        To dynamically retrieve online knowledge within a <a>N3 rule</a>,
```
Comment (TallTed) [inline spec/index.html:1194]:
```suggestion
        To dynamically retrieve online knowledge within an <a>N3 rule</a>,
```

File: spec/index.html [hunk 10]
```diff
@@ -2169,7 +1992,7 @@ <h4>N-ary Relations</h4>
       <p>Until now, we only considered binary relations between entities and/or values. But, many types of relations
         are
         <em>ternary, quaternary, or, in general, n-ary in nature</em>, i.e., they have an arbitrary number of
-        participants. Typical examples are purchase, employment, membership, .. relations.
+        participants. Typical examples are purchase, employment or membership relations.
```
Comment (TallTed) [inline spec/index.html:1954]:
```suggestion
        participants. Typical examples are purchase, employment, or membership relations.
```

**Comment by TallTed** [comment]
This is becoming the PR that ate the repository. 605 files changed, in 42 commits?!? It's all-but-impossible to review properly at that scale.

If the actual number of files intentionally being changed (e.g., actually excluding the umpteen `.DS_Store` that are in the list even though that's been added to `.gitignore` ... by this same PR) is small, maybe close this and start a new PR with just those (still) intended changes?

**Comment by domel** [comment]
I think that sec 4 (Useful Resources) looks more like a landing page rather than a W3C spec.

**Comment by william-vw** [comment]
Do you think we are ready to merge this PR - we discussed it last week, and, aside from some details, the CG seemed to be in agreement. I removed the "resources" section as requested. Other changes can be made later on in separate PRs.

**Review by domel — COMMENTED** [review]
duplicate ids:
basic-triples-example
object-lists-example
content_example

stragne attributes: 
cm-not-content
cm-text
(do we need it? If yes, please use data-cm-not-content)

A `p` end tag is in line 1345, but there is no open tag before.

**Comment by william-vw** [comment]
@domel 

> duplicate ids: `basic-triples-example` `object-lists-example` `content_example`
> 
> A `p` end tag is in line 1345, but there is no open tag before.

Fixed now by latest commits

> strange attributes: `cm-not-content` `cm-text` (do we need it? If yes, please use data-cm-not-content)
 
These are attributes added by CodeMirror (code sample styling) so yes, we need them

File: spec/index.html [hunk 11]
```diff
@@ -714,13 +717,13 @@ <h4>Blank node use cases</h4>
...
+    <section id='collections' class='informative'>
       <h3>Collections</h3>
       <p>We often need to describe ordered collections of things, e.g., a book written by several authors,
-        listing the students in a course, or the software modules within a package. N3 provides a succinct <dfn
-          data-lt="n3 collection|n3 list|list">collection</dfn> syntax to represent ordered collections of
-        resources enclosed by <code>(</code> <code>)</code>. The contained resources are called <dfn
-          data-lt="collection member">members</dfn>.</p>
+        listing the students in a course, or the software modules within a package. N3 supports <dfn
+          data-lt="n3 collection|n3 list|list">collections</dfn> as first-class citizens
+        to represent ordered collections of resources enclosed by <code>(</code> <code>)</code>.
```
Comment (TallTed) [inline spec/index.html:725]:
```suggestion
        to represent ordered collections of resources enclosed by parentheses 
        ("<code>(</code> <code>)</code>").
```

**Review by domel — APPROVED** [review]

**Comment by gkellogg** [comment]
> This is becoming the PR that ate the repository. 605 files changed, in 42 commits?!? It's all-but-impossible to review properly at that scale.
> 
> If the actual number of files intentionally being changed (e.g., actually excluding the umpteen `.DS_Store` that are in the list even though that's been added to `.gitignore` ... by this same PR) is small, maybe close this and start a new PR with just those (still) intended changes?

Most of it is adding code mirror to the library, and removing some remaining .DS_Store files.

But, in general, more frequent PRs with fewer changes stacked up is a good policy.

**Comment by william-vw** [comment]
Ok - looks like most folks in the CG agree with merging the PR (reached out to some over email; others had ample time to comment). And, it is true that this PR has dragged on for a while :-) So, merging it now.

**Pull request merged**